src/parser.ts

```diff
@@ -662,6 +672,165 @@ const TRAVERSE_BUILDER: MarkVisitor<(rhs: TraversalResult | null) => TraversalRe
   },
 }
 
+const SELECTOR_BUILDER: MarkVisitor<null> = {
+  square_bracket(p) {
+    p.process(TRAVERSE_BUILDER)
+    return null
+  },
+
+  slice(p) {
+    p.process(TRAVERSE_BUILDER)
+    return null
+  },
+
+  group(p) {
+    // This should only throw an error until we add support for groups in selectors.
+    throw new Error('Invalid selector syntax')
+  },
+
+  everything() {
+    throw new Error('Invalid selector syntax')
+  },
+
+  this() {
+    throw new Error('Invalid selector syntax')
+  },
+
+  parent() {
+    throw new Error('Invalid selector syntax')
+  },
+
+  dblparent(p) {
+    throw new Error('Invalid selector syntax')
+  },
+
+  traverse(p) {
+    // p.unshift()
+    const _ = p.process(EXPR_BUILDER)
+    p.process(TRAVERSE_BUILDER)
+    return null
+  },
```

Comment:
Two things here:

1. Right now it only processes the first traversal entry and will leave the parser in a weird state. We need to loop over until we find the `traversal_end`.
2. The base should be parsed using the selector builder to validate it properly.

```suggestion
  traverse(p) {
    p.process(SELECTOR_BUILDER)
    while (p.getMark().name !== 'traversal_end') {
      // TODO: This now starts parsing traversals as regular expressions.
      // We need a special builder for handling just the subset for selectors.
      p.process(TRAVERSE_BUILDER)
    }
    p.shift()
    return null
  },
```